### PR TITLE
#107: refactor callback concept

### PR DESCRIPTION
**Pull request by jarzynp**
#### Summary
Error and warning callbacks refactored. 
**Breaking change**

#### Description
Added old/new resource and updateActions (in case of error). Also some generic exception instead of separate message and exception.

#### Relevant Issues
https://github.com/commercetools/commercetools-sync-java/issues/107

### CODE REVIEW

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java [hunk 1]
```diff
@@ -144,7 +146,18 @@
                 final String errorMessage = format(CUSTOM_FIELDS_UPDATE_ACTIONS_BUILD_FAILED,
                     resourceTypeIdGetter.apply(oldResource),
                     resourceIdGetter.apply(oldResource), exception.getMessage());
-                syncOptions.applyErrorCallback(errorMessage, exception);
+
+                if (((ParameterizedType) syncOptions.getClass().getGenericSuperclass())
+                    .getActualTypeArguments()[0].getClass().isInstance(oldResource)) {
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: Why we need this check?

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java [hunk 2]
```diff
@@ -144,7 +146,18 @@
...
+                    syncOptions
+                        .applyErrorCallback(new SyncException(errorMessage, exception), oldResource,
+                            newResource,
+                            null);
+                } else {
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: What is the use case when it could happen?

File: src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+public interface QuadriConsumer<T, U, S, V> {
+
+    /**
+     * Performs operation on the given arguments.
+     *
+     * @param firstParam the first argument.
+     * @param secondParam the second argument.
+     * @param thirdParam the third argument.
+     * @param fourthParam the fourth argument.
+     */
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java]: I am not sure if such Java Doc is useful as it does not provide more information than its signature.

File: src/main/java/com/commercetools/sync/inventories/InventorySync.java [hunk 4]
```diff
@@ -281,7 +282,7 @@ private int calculateNumberOfBatches(int listSize) {
                     final Reference<Channel> supplyChannel = draft.getSupplyChannel();
                     final String errorMessage = format(CTP_INVENTORY_ENTRY_UPDATE_FAILED, draft.getSku(),
                         supplyChannel != null ? supplyChannel.getId() : null);
-                    handleError(errorMessage, exception, 1);
+                    handleError(errorMessage, exception, 1, entry, draft, updateActions);
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/inventories/InventorySync.java]: Maybe we should use here proper naming conventions too: entry -> oldInventory, draft -> newInventory? It is ok if we rather do in another PR.

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 5]
```diff
@@ -305,8 +306,8 @@
                 } catch (final BuildUpdateActionException buildUpdateActionException) {
                     final String errorMessage = format(FAILED_TO_BUILD_ATTRIBUTE_UPDATE_ACTION, attribute.getName(),
                         newProductVariant.getKey(), productKey, buildUpdateActionException.getMessage());
-                    syncOptions.applyErrorCallback(errorMessage,
-                        new BuildUpdateActionException(errorMessage, buildUpdateActionException));
+                    syncOptions.applyErrorCallback(new SyncException(errorMessage,
+                        new BuildUpdateActionException(errorMessage, buildUpdateActionException)), null, null, null);
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: Should not we pass here old and new resource too? 
@heshamMassoud Did not we have a convention to call warningCallback on sync validation and errorCallback on API errors only?

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 6]
```diff
@@ -318,7 +319,8 @@
             if (newAttribute == null) {
                 final String errorMessage = format(FAILED_TO_BUILD_ATTRIBUTE_UPDATE_ACTION, null,
                     newProductVariant.getKey(), productKey, NULL_PRODUCT_VARIANT_ATTRIBUTE);
-                syncOptions.applyErrorCallback(errorMessage, new BuildUpdateActionException(errorMessage));
+                syncOptions.applyErrorCallback(new SyncException(errorMessage,
+                    new BuildUpdateActionException(errorMessage)), null, null, updateActions);
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: I know that current method does not have whole old and new resource objects but maybe we should consider to provide those here? It is also ok to put it into separate issue/PR.

File: src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java [hunk 7]
```diff
@@ -280,7 +306,8 @@ private void handleError(@Nonnull final String errorMessage, @Nullable final Thr
                             final String errorMessage =
                                 format(CTP_PRODUCT_TYPE_UPDATE_FAILED, newProductType.getKey(),
                                     sphereException.getMessage());
-                            handleError(errorMessage, sphereException, 1);
+                            handleError(errorMessage, sphereException, 1, oldProductType, newProductType,
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java]: sphereException -> exception?

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java [hunk 8]
```diff
@@ -116,7 +117,9 @@
         final String draftKey = keyMapper.apply(draft);
 
         if (isBlank(draftKey)) {
-            syncOptions.applyErrorCallback(format(CREATE_FAILED, draftKey, "Draft key is blank!"));
+            syncOptions.applyErrorCallback(
```
Comment (butenkor) [inline src/main/java/com/commercetools/sync/services/impl/BaseService.java]: @heshamMassoud Does it make sense to call applyErrorCallback on sync validations? Especially when for similar issue we call warningCallback: https://github.com/commercetools/commercetools-sync-java/pull/367/files#diff-38b71d2c035833260351074420bf8067R60

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 9]
```diff
@@ -697,7 +719,9 @@ private void handleError(@Nonnull final String errorMessage,
                              @Nullable final Throwable exception,
                              final int failedTimes) {
 
-        syncOptions.applyErrorCallback(errorMessage, exception);
+        SyncException syncException = exception != null ? new SyncException(errorMessage, exception)
+            : new SyncException(errorMessage);
+        syncOptions.applyErrorCallback(syncException, null, null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: This is called in two cases actually: when `categoryService.cacheKeysToIds()` or `categoryService.fetchMatchingCategoriesByKeys(categoryKeysToFetch)` end in failure.
So this is some kind of preparation phase - getting data needed for further processing - there is no specific category yet. I think it has to stay this way unless you have another idea.

File: src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java [hunk 10]
```diff
@@ -80,11 +84,13 @@ private void clearSyncTestCollections() {
     }
 
     private ProductSyncOptions buildSyncOptions() {
-        final BiConsumer<String, Throwable> errorCallBack = (errorMessage, exception) -> {
-            errorCallBackMessages.add(errorMessage);
-            errorCallBackExceptions.add(exception);
-        };
-        final Consumer<String> warningCallBack = warningMessage -> warningCallBackMessages.add(warningMessage);
+        final QuadriConsumer<SyncException, Product, ProductDraft, Optional<List<UpdateAction<Product>>>>
+            errorCallBack = (exception, oldResource, newResource, updateActions) -> {
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java]: Will change this

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 11]
```diff
@@ -623,7 +625,7 @@ static boolean requiresChangeParentUpdateAction(@Nonnull final Category category
                                           () -> {
                                               if (!processedCategoryKeys.contains(categoryKey)) {
                                                   handleError(format(UPDATE_FAILED, categoryKey, sphereException),
-                                                      sphereException);
+                                                      sphereException, category, newCategory, updateActions);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: Will change this

File: src/main/java/com/commercetools/sync/categories/CategorySync.java [hunk 12]
```diff
@@ -680,7 +682,27 @@ static boolean requiresChangeParentUpdateAction(@Nonnull final Category category
...
+     * incrementing the total number of failed categories to sync.
+     *
+     * @param errorMessage The error message describing the reason(s) of failure.
+     * @param exception    The exception that called caused the failure, if any.
+     * @param category      the category to update.
+     * @param newCategory the category draft where we get the new data.
+     * @param updateActions the list of update actions to update the category with.
+     */
+    private void handleError(@Nonnull final String errorMessage, @Nullable final Throwable exception,
+        @Nullable final Category category,
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: Will change this

File: src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java [hunk 13]
```diff
@@ -252,8 +255,9 @@
                 newCategory.getAssets(),
                 new CategoryAssetActionFactory(syncOptions));
         } catch (final BuildUpdateActionException exception) {
-            syncOptions.applyErrorCallback(format("Failed to build update actions for the assets "
-                + "of the category with the key '%s'. Reason: %s", oldCategory.getKey(), exception), exception);
+            syncOptions.applyErrorCallback(new SyncException(format("Failed to build update actions for the assets "
+                + "of the category with the key '%s'. Reason: %s", oldCategory.getKey(), exception), exception),
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java]: You're completely right. Will remove this part.

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 14]
```diff
@@ -96,23 +104,25 @@ public void applyWarningCallback(@Nonnull final String warningMessage) {
...
+     * @param oldResource the old resource that is being compared to the new draft.
+     * @param newResourceDraft the new resource draft that is being compared to the old resource.
+     * @param updateActions the list of update actions.
      */
-    public void applyErrorCallback(@Nonnull final String errorMessage, @Nullable final Throwable exception) {
+    public void applyErrorCallback(@Nonnull final SyncException exception, @Nullable final U oldResource,
+        @Nullable final V newResourceDraft, @Nullable final List<UpdateAction<U>> updateActions) {
         if (this.errorCallBack != null) {
-            this.errorCallBack.accept(errorMessage, exception);
+            this.errorCallBack.accept(exception, oldResource, newResourceDraft, Optional.ofNullable(updateActions));
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Good point. I think all parameters except exception should be optional. Same goes with warningCallback. Do you agree?

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 15]
```diff
@@ -96,23 +104,25 @@ public void applyWarningCallback(@Nonnull final String warningMessage) {
...
         }
     }
 
     /**
      *
-     * @param errorMessage the error message to supply as first param to the {@code errorCallback} function.
-     * @see #applyErrorCallback(String, Throwable) applyErrorCallback(String, Throwable)
+     * @param errorMessage the error message to supply as part of first param to the {@code errorCallback} function.
+     * @see #applyErrorCallback(SyncException, Object, Object, List)
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: This was generated by IDE when I chose proper method. If I change to what you are proposing IDE is complaining: _Declaration has problems in Javadoc references_ and _cannot resolve symbol 'oldResource'_

We can either leave it like it is or change (it doesn't break generating of javadoc) to:
`@see #applyErrorCallback(SyncException exception, Object oldResource, Object newResource, List updateActions)`.
So what do you think?

File: src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java (hunk 3 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java]: I just followed pattern used in `TriConsumer` and `BiConsumer`

File: src/main/java/com/commercetools/sync/inventories/InventorySync.java (hunk 4 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/inventories/InventorySync.java]: Will change this

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 6 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: It is possible to provide old/new resource. Will add them.

File: src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java (hunk 7 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java]: Will change it.

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java (hunk 1 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: Old/new resource can be different than declared in syncOptions. F.e: 
`ProductSyncOptions extends `BaseSyncOptions<Product, ProductDraft>` so callbacks are expecting `Product` type as oldResource and `ProductDraft` as newResource. But there are situations where they are different: `Asset`/`AssetDraft`, `ProductVariant`/`ProductVariantDraft` etc. I will add comment in source code to clarify this.

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java (hunk 2 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r249038697

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 16]
```diff
@@ -96,23 +104,25 @@ public void applyWarningCallback(@Nonnull final String warningMessage) {
...
      */
     public void applyErrorCallback(@Nonnull final String errorMessage) {
-        applyErrorCallback(errorMessage, null);
+        applyErrorCallback(new SyncException(errorMessage), null, null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Currently:
- validating batch of drafts (`BatchProcessor#validateBatch`)
- preparing updateActions (`GenericUpdateActionUtils#buildTypedSetCustomTypeUpdateAction`,  `CustomUpdateActionUtils#buildCustomUpdateActions` and `ProductVariantUpdateActionUtils#buildProductVariantPricesUpdateActions`)

but:
- with providing old/new resource for https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r249037130 case `ProductVariantUpdateActionUtils#buildProductVariantPricesUpdateActions` will no longer be valid - will change this. 
- usage in `CustomUpdateActionUtils#buildCustomUpdateActions` can also be modified to same as for https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r249038697 - if condition is true provide old/new resource otherwise null

**Comment by jarzynp** [comment]
> It might be that not all requested changes fit in into this PR. In this case please create separate issue and link it to the comment.
@butenkor what did I miss?

File: src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java]: Would be cool if you could link commit which addresses the comment before resolving it.

File: src/main/java/com/commercetools/sync/categories/CategorySync.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: I see. So thats why you suggested below to make all arguments besides exception as Optional?

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (butenkor) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Last option is fine too. Thanks.

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java (hunk 1 above)
Comment (butenkor) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: It does not seem right that we would not pass anything in case of Assets issue besides the message. Is there a test for that? Also how could it be `ProductVariant/ProductVariantDraft` if those use product type attributes and not custom fields?

**Comment by butenkor** [comment]
> Did I miss something?

No you did not 👍 I just wanted to stress out that we might address some comments in different PRs if you feel that current PR might get to big.

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java (hunk 1 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: Tests:
- CustomUpdateActionUtilsTest, ProductVariantAssetUpdateActionUtilsTest -are using `Asset`
- BuildProductVariantPricesUpdateActionsTest - is using `ProductVariant`

File: src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java (hunk 7 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 6 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/main/java/com/commercetools/sync/inventories/InventorySync.java (hunk 4 above)
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/inventories/InventorySync.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/main/java/com/commercetools/sync/categories/CategorySync.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/categories/CategorySync.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/benchmark/java/com/commercetools/sync/benchmark/ProductSyncBenchmark.java]: Fixed: https://github.com/commercetools/commercetools-sync-java/pull/367/commits/70ca3bafc4e4a6a74e187f87cdcf2c38f3dfecf5

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 17]
```diff
@@ -22,19 +25,22 @@
...
-    protected BaseSyncOptions(@Nonnull final SphereClient ctpClient,
-                              @Nullable final BiConsumer<String, Throwable> errorCallBack,
-                              @Nullable final Consumer<String> warningCallBack,
-                              final int batchSize,
-                              @Nullable final TriFunction<List<UpdateAction<U>>, V, U, List<UpdateAction<U>>>
-                                  beforeUpdateCallback,
-                              @Nullable final Function<V, V> beforeCreateCallback) {
+    protected BaseSyncOptions(
+        @Nonnull final SphereClient ctpClient,
+        @Nullable final QuadriConsumer<SyncException, Optional<U>, Optional<V>, Optional<List<UpdateAction<U>>>>
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Good point, will change it.

File: src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+ * whose functional method is {@link #accept(Object, Object, Object, Object)}.
+ *
+ * @param <T> the type of the first argument to the function
+ * @param <U> the type of the second argument to the function
+ * @param <S> the type of the third argument to the function
+ * @param <V> the type of the fourth argument to the function
+ * @see Consumer
+ */
+@FunctionalInterface
+public interface QuadriConsumer<T, U, S, V> {
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/QuadriConsumer.java]: 🤦‍♂ and I was trying to find some good, meaningful name which go along Bi-, Tri- Consumer
`QuadConsumer` is much better, thanks! Will rename it.

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Totally forgot about this discussion. Will change order to match `beforeUpdateCallback` - same goes with order of parameters for warningCallback.
Should I also rename: `errorCallBack` to `errorCallback` and `warningCallBack` to `warningCallback` (property and method names)?

Example: `categoryService.cacheKeysToIds()`. There are more situation like this.
More details: https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r246772603 and https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r246749132

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: I think definitely it does, but I'm not sure it should be part of this PR

File: src/main/java/com/commercetools/sync/inventories/InventorySyncOptions.java [hunk 19]
```diff
@@ -10,21 +13,23 @@
...
-                         boolean ensureChannels,
-                         @Nullable final TriFunction<List<UpdateAction<InventoryEntry>>, InventoryEntryDraft,
-                             InventoryEntry, List<UpdateAction<InventoryEntry>>> beforeUpdateCallback,
-                         @Nullable final Function<InventoryEntryDraft, InventoryEntryDraft> beforeCreateCallback) {
+    InventorySyncOptions(
+        @Nonnull final SphereClient ctpClient,
+        @Nullable final QuadriConsumer<SyncException, Optional<InventoryEntry>, Optional<InventoryEntryDraft>,
+            Optional<List<UpdateAction<InventoryEntry>>>> updateActionErrorCallBack,
+        @Nullable final TriConsumer<SyncException, Optional<InventoryEntry>, Optional<InventoryEntryDraft>>
+            updateActionWarningCallBack,
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/inventories/InventorySyncOptions.java]: Ok, will change it.

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java [hunk 20]
```diff
@@ -22,19 +25,22 @@
...
+            errorCallBack,
+        @Nullable final TriConsumer<SyncException, Optional<U>, Optional<V>> warningCallBack,
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: Consistency reasons, more details in second part of https://github.com/commercetools/commercetools-sync-java/pull/367#discussion_r286514838

File: src/main/java/com/commercetools/sync/inventories/InventorySync.java [hunk 21]
```diff
@@ -324,7 +326,29 @@ private int calculateNumberOfBatches(int listSize) {
      */
     private void handleError(@Nonnull final String errorMessage, @Nullable final Throwable exception,
                              final int failedTimes) {
-        syncOptions.applyErrorCallback(errorMessage, exception);
+        SyncException syncException = exception != null ? new SyncException(errorMessage, exception)
+            : new SyncException(errorMessage);
+        syncOptions.applyErrorCallback(syncException, null, null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/inventories/InventorySync.java]: I think I don't fully understand what you mean. 
There are two kinds of method `handleError` One is called when we don't have new resource or draft (f.e: `InventoryEntry#fetchExistingInventories`), and second when those values are known (f.e: `InventoryEntry#create`). And the second one is being called in situation you described.

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantPriceUpdateActionUtils.java [hunk 22]
```diff
@@ -19,19 +20,22 @@
...
+     * {@link UpdateAction}&lt;{@link Product}&gt; as a result. If both the {@link Price} and the
+     * {@link PriceDraft} have identical fields, then no update action is needed and hence an empty
+     * {@link List} is returned.
      *
-     * @param variantId   the variantId needed for building the update action.
-     * @param oldPrice    the price which should be updated.
-     * @param newPrice    the price draft where we get the new fields.
-     * @param syncOptions responsible for supplying the sync options to the sync utility method. It is used for
-     *                    triggering the error callback within the utility, in case of errors.
+     * @param variantId the variantId needed for building the update action.
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantPriceUpdateActionUtils.java]: My bad, I tried to configure my IDEA to avoid it, but without clear formatting guidelines it is quite hard. Will revert this change.

File: src/main/java/com/commercetools/sync/services/impl/CustomerGroupServiceImpl.java [hunk 23]
```diff
@@ -54,7 +56,10 @@ public CustomerGroupServiceImpl(@Nonnull final BaseSyncOptions syncOptions) {
                 if (isNotBlank(fetchedCustomerGroupKey)) {
                     keyToIdCache.put(fetchedCustomerGroupKey, fetchedCustomerGroupId);
                 } else {
-                    syncOptions.applyWarningCallback(format(CUSTOMER_GROUP_KEY_NOT_SET, fetchedCustomerGroupId));
+                    //syncOptions is ProductSyncOptions
+                    syncOptions.applyWarningCallback(
+                        new SyncException(format(CUSTOMER_GROUP_KEY_NOT_SET, fetchedCustomerGroupId)),
+                        null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/services/impl/CustomerGroupServiceImpl.java]: Because we don't have it here. It fetches list of `CustomerGroup` and it is using `ProductSyncOptions`. Types don't match.

File: src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java [hunk 24]
```diff
@@ -61,8 +63,10 @@ public ProductTypeServiceImpl(@Nonnull final BaseSyncOptions syncOptions) {
                     if (StringUtils.isNotBlank(fetchedTypeKey)) {
                         keyToIdCache.put(fetchedTypeKey, id);
                     } else {
-                        syncOptions.applyWarningCallback(format("ProductType with id: '%s' has no key set. Keys are"
-                                + " required for productType matching.", id));
+                        // syncOptions is ProductSyncOptions
+                        syncOptions.applyWarningCallback(
+                            new SyncException(format("ProductType with id: '%s' has no key set. Keys are"
+                                + " required for productType matching.", id)), null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java]: Because we don't have it here. It fetches list of `ProductType` and it is using `ProductSyncOptions`. Types don't match (`ProductType` is not `Product`).

File: src/main/java/com/commercetools/sync/services/impl/TaxCategoryServiceImpl.java [hunk 25]
```diff
@@ -49,8 +50,9 @@ public TaxCategoryServiceImpl(@Nonnull final ProductSyncOptions syncOptions) {
                 if (StringUtils.isNotBlank(fetchedTaxCategoryKey)) {
                     keyToIdCache.put(fetchedTaxCategoryKey, id);
                 } else {
-                    syncOptions.applyWarningCallback(format("TaxCategory with id: '%s' has no key set. Keys are"
-                        + " required for taxCategory matching.", id));
+                    syncOptions.applyWarningCallback(
+                        new SyncException(format("TaxCategory with id: '%s' has no key set. Keys are"
+                        + " required for taxCategory matching.", id)), null, null);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/services/impl/TaxCategoryServiceImpl.java]: Because we don't have it here. It fetches list of `TaxCategory` and it is using `ProductSyncOptions`. Types don't match (`TaxCategory` is not `Product`).

File: src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java [hunk 26]
```diff
@@ -133,6 +136,12 @@
         @Nonnull final Function<T, String> updateIdGetter,
         @Nonnull final BaseSyncOptions syncOptions) {
 
+        // oldResource/newResource types can be different than declared in syncOptions, thus can't
+        // be used in errorCallback; example:
+        // ProductSyncOptions<Product, ProductDraft> but oldResource/newResource can be Asset/AssetDraft
+        boolean resourceMatchWithSyncOptions = ((ParameterizedType) syncOptions.getClass().getGenericSuperclass())
+            .getActualTypeArguments()[0].getClass().isInstance(oldResource);
```
[the remark this answers is not in the document]
Comment (jarzynp) [reply src/main/java/com/commercetools/sync/commons/utils/CustomUpdateActionUtils.java]: What exactly do you mean by _skipping this logic_?
We can always put `null` there, but then we lose the working solution in case the argument types of syncOptions and resource match each other.

Maybe better option is sth like this?
add new parameter to BaseSyncOptions -> `Pair<Class<V>, Class<U>> types` and a method, sth like `public boolean <T extends Custom, S extends CustomDraft> matchingTypes(T oldResource, S newResource)` which will do simple check:
`return types.getKey().isAssignableFrom(newResource.getClass()) && types.getValue()).isAssignableFrom(oldResource.getClass())`

Than we can replace this hackish solution to more future-proof. I'm not saying it is ideal or elegant solution but at least it should somehow solve current problem.

**Comment by heshamMassoud** [comment]
Hi @jarzynp, there has been many changes on the master after this PR. 

If you don't mind me, I will make this PR up-to-date with master and then address/resolve all the comments on it, to make it ready for merging :)

Are you fine with that?

**Comment by jarzynp** [comment]
Hi
Sure, go ahead
I just pushed some changes I forget to commit earlier.

File: src/main/java/com/commercetools/sync/services/impl/BaseService.java (hunk 8 above)
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/impl/BaseService.java:121]: Hi @butenkor is the reply from @heshamMassoud clear or are there still points that need to be checked?

File: src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/commons/BaseSyncOptions.java]: @butenkor I think it's enough that we marked it with nullable, right? Or do you think we should now rather use Optional in these cases? Right now the method signature is this: `void accept(@Nullable final T firstParam, @Nullable final U secondParam, @Nullable final V thirdParam);`. If we decide to use Optional, then I guess Nullable is obsolete, and we change it to `void accept(@Nullable final T firstParam, final Optional<U> secondParam, @Nullable final Optional<V> thirdParam);`

File: src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java:69]: I guess in this case we can pass `productType` to the callback as `oldResource` param.

File: src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java [hunk 27]
```diff
@@ -61,8 +63,10 @@ public ProductTypeServiceImpl(@Nonnull final BaseSyncOptions syncOptions) {
                     if (StringUtils.isNotBlank(fetchedTypeKey)) {
                         keyToIdCache.put(fetchedTypeKey, id);
                     } else {
-                        syncOptions.applyWarningCallback(format("ProductType with id: '%s' has no key set. Keys are"
-                                + " required for productType matching.", id));
+                        // syncOptions is ProductSyncOptions
```
Comment (lojzatran) [inline src/main/java/com/commercetools/sync/services/impl/ProductTypeServiceImpl.java:66]: This comment doesn't seem to be valid. I had a look at an integration test which uses this class and SyncOptionc there is of type `ProductTypeSyncOptions`

File: src/main/java/com/commercetools/sync/services/impl/CustomerGroupServiceImpl.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/services/impl/CustomerGroupServiceImpl.java:62]: @butenkor What do you think is the solution here? Should we leave `null` or create a CustomerGroupSyncOptions and pass there the failed resources?

File: src/main/java/com/commercetools/sync/inventories/InventorySync.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/inventories/InventorySync.java:331]: `InventorySync.java:validateDraft()` we could pass `inventoryDraft` to error callback in case SKU is null. Maybe there are more cases like this where we. Same applies for `InventorySync.java:syncBatch()`.

**Comment by ahmetoz** [comment]
PR will continue in #521
